networking/src/NTP/Client.hs

```diff
@@ -280,22 +277,19 @@ stopNtpClient cli = do
     -- unblock receiving from socket in case no one replies
     forM_ sockets $ \s -> liftIO (close s) `catchAny` (const $ pure ())
 
-startNtpClient :: NtpMonad m => NtpClientSettings m -> m (NtpStopButton m)
-startNtpClient settings =
+withNtpClient :: NtpMonad m => NtpClientSettings m -> (NtpStopButton m -> m a) -> m a
+withNtpClient settings callback =
     withSocketsDoLifted $
     bracketOnError (mkSockets settings) closeSockets $ \sock -> do
         cli <- mkNtpClient settings sock
 
         addrs <- catMaybes <$> mapM (resolveHost cli $ socketsToBoolDescr sock)
                                     (ntpServers settings)
-        if null addrs then
-            throwM NoHostResolved
-        else do
-            void . fork . withSocketsDoLifted $ startReceive cli
-            void . fork . withSocketsDoLifted $ startSend addrs cli
-            log cli Info "Launched"
-
-        return $ NtpStopButton $ stopNtpClient cli
+        when (null addrs) $ throwM NoHostResolved
+        withAsync (withSocketsDoLifted $ startReceive cli) $ \_ ->
+            withAsync (withSocketsDoLifted $ startSend addrs cli) $ \_ -> do
+                log cli Info "Launched NTP client"
+                callback (NtpStopButton $ stopNtpClient cli)
```

Comment:
Since it's now written this way, it seems better to press stop button here (which will close sockets) rather than export button outside

Reply:
Also, this pretty code
https://github.com/input-output-hk/cardano-sl/blob/d9e2fd91bc9e1a7b8a3d933c55c5a5f37c32c6dd/networking/src/NTP/Client.hs#L256:L259
becomes unsafe because exception can fly in during `=<<` and created socket will remain unused :confused:

Reply:
> Also, this pretty code becomes unsafe

Was it safe before? To be honest the whole design of `NTP.Client` looks not perfect, but in this PR I only wanted to spawn threads properly.